pilot/cmd/pilot-agent/status/server.go

```diff
@@ -387,42 +388,67 @@ type PrometheusScrapeConfiguration struct {
 // The merge here is a simple string concatenation. This works for almost all cases, assuming the application
 // is not exposing the same metrics as Envoy.
+// This merging works for both FmtText and FmtOpenMetrics and will use the format of the application metrics
 // Note that we do not return any errors here. If we do, we will drop metrics. For example, the app may be having issues,
 // but we still want Envoy metrics. Instead, errors are tracked in the failed scrape metrics/logs.
 func (s *Server) handleStats(w http.ResponseWriter, r *http.Request) {
 	metrics.ScrapeTotals.Increment()
 	var envoy, application, agent []byte
 	var err error
 	// Gather all the metrics we will merge
-	if envoy, err = s.scrape(fmt.Sprintf("http://localhost:%d/stats/prometheus", s.envoyStatsPort), r.Header); err != nil {
+	if envoy, _, err = s.scrape(fmt.Sprintf("http://localhost:%d/stats/prometheus", s.envoyStatsPort), r.Header); err != nil {
 		log.Errorf("failed scraping envoy metrics: %v", err)
 		metrics.EnvoyScrapeErrors.Increment()
 	}
+	// Process envoy's metrics to make them compatible with FmtOpenMetrics
+	envoy = processMetrics(envoy)
+
+	// Scrape app metrics if defined and capture their format
+	var format expfmt.Format
 	if s.prometheus != nil {
+		var contentType string
 		url := fmt.Sprintf("http://localhost:%s%s", s.prometheus.Port, s.prometheus.Path)
-		if application, err = s.scrape(url, r.Header); err != nil {
+		if application, contentType, err = s.scrape(url, r.Header); err != nil {
 			log.Errorf("failed scraping application metrics: %v", err)
 			metrics.AppScrapeErrors.Increment()
 		}
+		format = negotiateMetricsFormat(contentType)
 	}
+
 	if agent, err = scrapeAgentMetrics(); err != nil {
 		log.Errorf("failed scraping agent metrics: %v", err)
 		metrics.AgentScrapeErrors.Increment()
 	}
 
+	w.Header().Set("Content-Type", string(format))
+
 	// Write out the metrics
+	if _, err := w.Write(agent); err != nil {
+		log.Errorf("failed to write agent metrics: %v", err)
+		metrics.AgentScrapeErrors.Increment()
+	}
 	if _, err := w.Write(envoy); err != nil {
 		log.Errorf("failed to write envoy metrics: %v", err)
 		metrics.EnvoyScrapeErrors.Increment()
 	}
+	// App metrics must go last because if they are FmtOpenMetrics,
+	// they will have a trailing "# EOF" which terminates the full exposition
 	if _, err := w.Write(application); err != nil {
 		log.Errorf("failed to write application metrics: %v", err)
 		metrics.AppScrapeErrors.Increment()
 	}
-	if _, err := w.Write(agent); err != nil {
-		log.Errorf("failed to write agent metrics: %v", err)
-		metrics.AgentScrapeErrors.Increment()
+}
+
+func negotiateMetricsFormat(contentType string) expfmt.Format {
+	mediaType, _, err := mime.ParseMediaType(contentType)
+	if err == nil && mediaType == expfmt.OpenMetricsType {
+		return expfmt.FmtOpenMetrics
 	}
+	return expfmt.FmtText
+}
+
+func processMetrics(metrics []byte) []byte {
+	return []byte(strings.ReplaceAll(string(metrics), "\n\n", "\n"))
```

Comment:
very nit: I think you can do bytes.ReplaceAll as well to avoid double conversion

Reply:
great call! updated